# Starter code

In [31]:
import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup
import re
import json
import time 
from datetime import date
import csv

In [2]:
playerLink = 'https://www.pro-football-reference.com/players/W/WalkTr03.htm'

In [3]:
base = 'https://www.pro-football-reference.com'

In [28]:
playerDF = pd.DataFrame(columns = ['draft_year', 'draft_round' , 'draft_pick', 'draft_team', 'player_name', 'player_url', 'position', 'draft_age','college', 'height', 'weight', 'birth_year', 'birth_month', 'birth_day', 'college_stats_url', 'birth_place', 'high_school', 'image_url'])

In [61]:
playerDF

draft_year draft_round draft_pick draft_team         player_name  \
0          2022           1          1        JAX       Travon Walker   
1          2022           1          2        DET    Aidan Hutchinson   
2          2022           1          3        HOU  Derek Stingley Jr.   
3          2022           1          4        NYJ       Sauce Gardner   
4          2022           1          5        NYG   Kayvon Thibodeaux   
...         ...         ...        ...        ...                 ...   
7134       1994           1         27        BUF         Jeff Burris   
7135       1994           1         28        SFO       William Floyd   
7136       1994           1         29        CLE   Derrick Alexander   
7137       1994           2         30        CIN        Darnay Scott   
7138       1994           2         31        WAS        Tre' Johnson   

                                             player_url position draft_age  \
0     https://www.pro-football-reference.com/players...       DE        21   
1     https://www.pro-football-reference.com/players...       DE        22   
2     https://www.pro-football-reference.com/players...       CB        21   
3     https://www.pro-football-reference.com/players...       CB        22   
4     https://www.pro-football-reference.com/players...       DE        21   
...                                                 ...      ...       ...   
7134  https://www.pro-football-reference.com/players...       DB        22   
7135  https://www.pro-football-reference.com/players...       FB        22   
7136  https://www.pro-football-reference.com/players...       WR        22   
7137  https://www.pro-football-reference.com/players...       WR        22   
7138  https://www.pro-football-reference.com/players...        G        23   

            college height   weight birth_year birth_month birth_day  \
0           Georgia    6-5  275 lbs       2000          12        18   
1          Michigan    6-7  265 lbs       2000          08        09   
2               LSU    6-1  195 lbs       2001          06        20   
3        Cincinnati    6-3  200 lbs       2000          08        31   
4            Oregon    6-5  250 lbs       2000          12        15   
...             ...    ...      ...        ...         ...       ...   
7134     Notre Dame    6-0  190 lbs       1972          06        07   
7135    Florida St.    6-1  242 lbs       1972          02        17   
7136       Michigan    6-2  195 lbs       1971          11        06   
7137  San Diego St.    6-1  200 lbs       1972          07        07   
7138         Temple    6-2  328 lbs       1971          08        30   

                                      college_stats_url  \
0     http://www.sports-reference.com/cfb/players/tr...   
1     http://www.sports-reference.com/cfb/players/ai...   
2     http://www.sports-reference.com/cfb/players/de...   
3     http://www.sports-reference.com/cfb/players/ah...   
4     http://www.sports-reference.com/cfb/players/ka...   
...                                                 ...   
7134  http://www.sports-reference.com/cfb/players/je...   
7135  http://www.sports-reference.com/cfb/players/wi...   
7136  http://www.sports-reference.com/cfb/players/de...   
7137  http://www.sports-reference.com/cfb/players/da...   
7138                                               None   

                  birth_place                                  high_school  \
0          Thomaston, GA, USA                                  HS,Upson,GA   
1                          MI                           HS,Divine Child,MI   
2        Baton Rouge, LA, USA                                 HS,Dunham,LA   
3            Detroit, MI, USA                     HS,Martin Luther King,MI   
4                          CA  HS,Susan Miller Dorsey,CA,Oaks Christian,CA   
...                       ...                                          ...   
7134       Rock Hill, SC, USA                           HS,Northwestern

# Trash but noteworthy to make progress, more for testing 


In [211]:
year = 2022

In [607]:
import warnings

with warnings.catch_warnings():
    warnings.simplefilter("ignore")

In [ ]:
# Still not sure how Michael found this, but it helps

text = playerSoup.find_all(text=re.compile("birth"))
text

['\n{\n    "@context": "http://schema.org",\n    "@type": "Person",\n    "name": "Travon Walker",\n    "url": "https://www.pro-football-reference.com/players/W/WalkTr03.htm",\n    "image": {\n        "@type": "ImageObject",\n        "caption": "Travon Walker, Jacksonville Jaguars", \n        "representativeOfPage": true, \n        "contentUrl": "https://www.pro-football-reference.com/req/20180910/images/headshots/WalkTr03_2022.jpg"\n    },\n    "memberOf": {\n        "@type": "SportsTeam",\n        "name": "Jacksonville Jaguars",\n        "sport": "Football"\n    },\n    "birthDate": "2000-12-18"\n    ,\n  "birthPlace": "Thomaston, GA, USA",\n    "height": { "@type": "QuantitativeValue", "value": "6-5" },\n    "weight": { "@type": "QuantitativeValue", "value": "275 lbs" }\n}\n']

In [ ]:
playerDict = playerDictionaryConv(text)
playerDict

{'@context': 'http://schema.org',
 '@type': 'Person',
 'name': 'Travon Walker',
 'url': 'https://www.pro-football-reference.com/players/W/WalkTr03.htm',
 'image': {'@type': 'ImageObject',
  'caption': 'Travon Walker, Jacksonville Jaguars',
  'representativeOfPage': True,
  'contentUrl': 'https://www.pro-football-reference.com/req/20180910/images/headshots/WalkTr03_2022.jpg'},
 'memberOf': {'@type': 'SportsTeam',
  'name': 'Jacksonville Jaguars',
  'sport': 'Football'},
 'birthDate': '2000-12-18',
 'birthPlace': 'Thomaston, GA, USA',
 'height': {'@type': 'QuantitativeValue', 'value': '6-5'},
 'weight': {'@type': 'QuantitativeValue', 'value': '275 lbs'}}

In [ ]:
getPersonalStats(playerLink)

{'@context': 'http://schema.org', '@type': 'Person', 'name': 'Travon Walker', 'url': 'https://www.pro-football-reference.com/players/W/WalkTr03.htm', 'image': {'@type': 'ImageObject', 'caption': 'Travon Walker, Jacksonville Jaguars', 'representativeOfPage': True, 'contentUrl': 'https://www.pro-football-reference.com/req/20180910/images/headshots/WalkTr03_2022.jpg'}, 'memberOf': {'@type': 'SportsTeam', 'name': 'Jacksonville Jaguars', 'sport': 'Football'}, 'birthDate': '2000-12-18', 'birthPlace': 'Thomaston, GA, USA', 'height': {'@type': 'QuantitativeValue', 'value': '6-5'}, 'weight': {'@type': 'QuantitativeValue', 'value': '275 lbs'}}


In [ ]:
draftPlayerDF = pd.DataFrame(columns = [['draft_year', 'draft_round' , 'draft_pick', 'draft_team', 'player_name', 'player_url', 'position', 'draft_age','college', 'height', 'weight', 'birth_year', 'birth_month', 'birth_day', 'college_stats_url', 'birth_place', 'image_url']])

In [ ]:
draft = requestDraftPage(2022)

In [ ]:
draftPlayers = getDraftPlayers(draft)[0:2]
draftPlayers

[<tr><th class="right " data-stat="draft_round" scope="row">1</th><td class="right " csk="0.5" data-stat="draft_pick">1</td><td class="left " csk="jax01" data-stat="team"><a href="/teams/jax/2022_draft.htm" title="Jacksonville Jaguars">JAX</a></td><td class="left " csk="Walker,Travon" data-append-csv="WalkTr03" data-stat="player"><strong><a href="/players/W/WalkTr03.htm">Travon Walker</a></strong></td><td class="left " csk="DE--------0001" data-stat="pos">DE</td><td class="right " data-stat="age">21</td><td class="right " data-stat="year_max">2022</td><td class="right iz" data-stat="all_pros_first_team">0</td><td class="right iz" data-stat="pro_bowls">0</td><td class="right iz" data-stat="years_as_primary_starter">0</td><td class="right " data-stat="career_av">6</td><td class="right " data-stat="draft_av">6</td><td class="right " data-stat="g">15</td><td class="right iz" data-stat="pass_cmp">0</td><td class="right iz" data-stat="pass_att">0</td><td class="right iz" data-stat="pass_yds"

In [ ]:
draftPlayers[0].find_all('a')[-1]['href']

'http://www.sports-reference.com/cfb/players/travon-walker-1.html'

In [ ]:
extractPlayerData(draftPlayers, 2022, base)

[<tr><th class="right " data-stat="draft_round" scope="row">1</th><td class="right " csk="0.5" data-stat="draft_pick">1</td><td class="left " csk="jax01" data-stat="team"><a href="/teams/jax/2022_draft.htm" title="Jacksonville Jaguars">JAX</a></td><td class="left " csk="Walker,Travon" data-append-csv="WalkTr03" data-stat="player"><strong><a href="/players/W/WalkTr03.htm">Travon Walker</a></strong></td><td class="left " csk="DE--------0001" data-stat="pos">DE</td><td class="right " data-stat="age">21</td><td class="right " data-stat="year_max">2022</td><td class="right iz" data-stat="all_pros_first_team">0</td><td class="right iz" data-stat="pro_bowls">0</td><td class="right iz" data-stat="years_as_primary_starter">0</td><td class="right " data-stat="career_av">6</td><td class="right " data-stat="draft_av">6</td><td class="right " data-stat="g">15</td><td class="right iz" data-stat="pass_cmp">0</td><td class="right iz" data-stat="pass_att">0</td><td class="right iz" data-stat="pass_yds"

TypeError: ignored

{'@context': 'http://schema.org',
 '@type': 'Person',
 'name': 'Travon Walker',
 'url': 'https://www.pro-football-reference.com/players/W/WalkTr03.htm',
 'image': {'@type': 'ImageObject',
  'caption': 'Travon Walker, Jacksonville Jaguars',
  'representativeOfPage': True,
  'contentUrl': 'https://www.pro-football-reference.com/req/20180910/images/headshots/WalkTr03_2022.jpg'},
 'memberOf': {'@type': 'SportsTeam',
  'name': 'Jacksonville Jaguars',
  'sport': 'Football'},
 'birthDate': '2000-12-18',
 'birthPlace': 'Thomaston, GA, USA',
 'height': {'@type': 'QuantitativeValue', 'value': '6-5'},
 'weight': {'@type': 'QuantitativeValue', 'value': '275 lbs'}}

In [ ]:
f = createEmptyPlayer()

In [ ]:
f

{'draft_year': None,
 'draft_round': None,
 'draft_pick': None,
 'draft_team': None,
 'player_name': None,
 'player_url': None,
 'position': None,
 'draft_age': None,
 'college': None,
 'height': None,
 'weight': None,
 'birth_year': None,
 'birth_month': None,
 'birth_day': None,
 'college_stats_url': None}

In [ ]:
draftPlayers

[<tr><th class="right " data-stat="draft_round" scope="row">1</th><td class="right " csk="0.5" data-stat="draft_pick">1</td><td class="left " csk="jax01" data-stat="team"><a href="/teams/jax/2022_draft.htm" title="Jacksonville Jaguars">JAX</a></td><td class="left " csk="Walker,Travon" data-append-csv="WalkTr03" data-stat="player"><strong><a href="/players/W/WalkTr03.htm">Travon Walker</a></strong></td><td class="left " csk="DE--------0001" data-stat="pos">DE</td><td class="right " data-stat="age">21</td><td class="right " data-stat="year_max">2022</td><td class="right iz" data-stat="all_pros_first_team">0</td><td class="right iz" data-stat="pro_bowls">0</td><td class="right iz" data-stat="years_as_primary_starter">0</td><td class="right " data-stat="career_av">6</td><td class="right " data-stat="draft_av">6</td><td class="right " data-stat="g">15</td><td class="right iz" data-stat="pass_cmp">0</td><td class="right iz" data-stat="pass_att">0</td><td class="right iz" data-stat="pass_yds"

In [ ]:
draftPlayers.find('td', {'data-stat': 'draft_pick'}).get_text()

AttributeError: ignored

# Lines for testing when errors come up

In [8]:
df1 = playerDF

In [50]:
year = 1999

In [51]:
# Test for specific year
dp = requestDraftPage(year)

In [56]:
# Grab individual players to test what happens
players = getDraftPlayers(dp)[174:]

In [53]:
len(players)

80

In [58]:
extractPlayerData(players,playerDF,year)

# The looping function

In [62]:
for year in np.arange(1994,1989,-1):
    
    dp = requestDraftPage(year)

    players = getDraftPlayers(dp)

    extractPlayerData(players,playerDF,year)

# DF to check

In [108]:
playerDF.tail(10)

draft_year draft_round draft_pick draft_team        player_name  \
234        2010           7        244        PHI       Kurt Coleman   
235        2010           7        245        SEA       Jameson Konz   
236        2010           7        247        NWE  Brandon Deaderick   
237        2010           7        249        CAR     Robert McClain   
238        2010           7        250        NWE       Zac Robinson   
239        2010           7        251        OAK       Stevie Brown   
240        2010           7        252        MIA     Austin Spitler   
241        2010           7        253        TAM         Erik Lorig   
242        2010           7        254        STL          Josh Hull   
243        2010           7        255        DET          Tim Toone   

                                            player_url position draft_age  \
234  https://www.pro-football-reference.com/players...       DB        22   
235  https://www.pro-football-reference.com/players...       TE        24   
236  https://www.pro-football-reference.com/players...       DE        23   
237  https://www.pro-football-reference.com/players...       DB        22   
238  https://www.pro-football-reference.com/players...       QB        23   
239  https://www.pro-football-reference.com/players...       DB        23   
240  https://www.pro-football-reference.com/players...       LB        23   
241  https://www.pro-football-reference.com/players...       DE        23   
242  https://www.pro-football-reference.com/players...       LB        23   
243  https://www.pro-football-reference.com/players...       WR        25   

          college height   weight birth_year birth_month birth_day  \
234      Ohio St.   5-11  208 lbs       1988          07        01   
235      Kent St.    6-3  227 lbs       1986          07        02   
236       Alabama    6-4  297 lbs       1987          08        19   
237   Connecticut    5-9  180 lbs       1988          07        22   
238  Oklahoma St.    6-3  218 lbs       1986          09        29   
239      Michigan    6-0  213 lbs       1987          07        17   
240      Ohio St.    6-3  244 lbs       1986          10        26   
241      Stanford    6-4  265 lbs       1986          11        17   
242      Penn St.    6-2  236 lbs       1987          05        21   
243     Weber St.   5-10  185 lbs       1985          02        14   

                                     college_stats_url  \
234  http://www.sports-reference.com/cfb/players/ku...   
235                                               None   
236  http://www.sports-reference.com/cfb/players/br...   
237  http://www.sports-reference.com/cfb/players/ro...   
238                                               None   
239  http://www.sports-reference.com/cfb/players/st...   
240  http://www.sports-reference.com/cfb/players/au...   
241  http://www.sports-reference.com/cfb/players/er...   
242  http://www.sports-reference.com/cfb/players/jo...   
243                                               None   

                birth_place                   high_school  \
234        Clayton, OH, USA               HS,Northmont,OH   
235      Uniontown, OH, USA                    HS,Lake,OH   
236  Elizabethtown, KY, USA           HS,Elizabethtown,KY   
237          Lusby, MD, USA                HS,Patuxent,MD   
238                    None                            HS   
239         Dallas, TX, USA           HS,Columbus East,IN   
240         Dayton, OH, USA               HS,Bellbrook,OH   
241  Rolling Hills, CA, USA  HS,Palos Verdes Peninsula,CA   
242  State College, PA, USA            HS,Penns Valley,PA   
243                    None                            HS   

                                             image_url  
234  https://www.pro-football-reference.com/req/201...  
235  https://www.pro-football-reference.com/req/201...  
236  https://www.pro-football-reference.com/req/201...  
237  https://www.pro-football-reference.com/req/201...

In [65]:
#playerDF.to_csv('draft_players_2022-2000.csv', index=False)

# RequestDraftPage(year) Function

In [34]:
def requestDraftPage(year):
  base = 'https://www.pro-football-reference.com'
  yearLink = f'/years/{year}/draft.htm'

  draftPage = requests.get(base+yearLink)

  return draftPage

# GetDraftPlayer Function

In [35]:
def getDraftPlayers(draftPage):
    soup = BeautifulSoup(draftPage.text, 'html.parser')

    draftPlayers = soup.find('tbody').find_all('tr', attrs={'class': None})

    return draftPlayers

# ExtractPlayerData Function

In [36]:
def extractPlayerData(playerList, df, year):
    
  nextIndex = len(df)

  for player in playerList:
        
    # Blank list that will take on values of each player
    playerRecord = createEmptyPlayer()

    # Function to retrieve data in the draft listing
    playerRecord = extractPlayerBasicDataMapping(playerRecord, player, year)
    
    
    # Function to retrieve data in the draft listing using the data-stat attribute 
    playerRecord = dataStatAttrsInsert(player, playerRecord)
    
    if playerRecord['player_url'] != None:
    # Calling function to get players height, weight and birthday located in their personal pages
        playerRecord = getPersonalStats(playerRecord['player_url'], playerRecord)
    
    saveToCSV(playerRecord, df)

    df.loc[nextIndex] = playerRecord
    nextIndex += 1


# ExtractPlayerData Interior Functions

In [37]:
def createEmptyPlayer():
    keys = ['draft_year', 'draft_round' , 'draft_pick', 'draft_team', 'player_name', 'player_url', 'position', 'draft_age','college', 'height', 'weight', 'birth_year', 'birth_month', 'birth_day', 'college_stats_url', 'birth_place', 'high_school', 'image_url']

    d = {key: None for key in keys}

    return d

In [38]:
def extractPlayerBasicDataMapping(playerRecord, player, year):

    try:
      playerRecord['draft_year'] = year
    except:
      playerRecord['draft_year'] = None 

    try:
        secondLink = player.find_all('a')[1]['href']
    except IndexError:
        playerRecord['player_url'] = None
        secondLink = ''
    
    if '/players' in  secondLink:  
        try:
            playerRecord['player_url'] = base+player.find_all('a')[1]['href']
        except:
            playerRecord['player_url'] = None 
    else:
        playerRecord['player_url'] = None 

    #if playerRecord['player_url'] != None:
    

      
      #try:
      #  playerRecord['draft_round'] = player.find('th').get_text()
      #except:
      #  playerRecord['draft_round'] = None

    
    return playerRecord

In [39]:
def dataStatAttrsInsert(playerPage, record): 
    
    playerDataStatList = ['draft_pick', 'player', 'pos', 'college_id', 'team', 'college_link', 'age', 'draft_round']
    for name in playerDataStatList:
        value = extract_value(name, playerPage)
        if name == 'draft_pick':
            record['draft_pick'] = value
        elif name == 'player':
            record['player_name'] = value
        elif name == 'pos':
            record['position'] = value
        elif name == 'team':
            record['draft_team'] = value
        elif name == 'college_link':
            record['college_stats_url'] = value
        elif name == 'age':
            record['draft_age'] = value
        elif name == 'draft_round':
            record['draft_round'] = value
        else:
            record['college'] = value
    return record

In [40]:
def saveToCSV(record, df):

    record = pd.DataFrame(record, index=[0])

    record.to_csv('nfl_draft_players_2022-1990.csv', mode = 'a', index=False, header = False)
    

# dataStatAttrsInsert interior function 

In [41]:
def extract_value(name, playerPage):
    
    if name == 'draft_pick':
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None

    elif name == 'draft_round':
        try:
            return playerPage.find('th', {'data-stat': name}).get_text()
        except AttributeError:
            return None

    elif name == 'player':
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None

    elif name == 'pos':
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None

    elif name == 'team':
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None

    elif name == 'college_link':
        td = playerPage.find('td', {'data-stat': name})
        if td:
            a = td.find('a')
            if a:
                return a['href']
            else:
                return None
        else:
            return None

    elif name == 'age':
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None
    else:
        try:
            return playerPage.find('td', {'data-stat': name}).get_text()
        except AttributeError:
            return None

# GetPersonalStats Function

In [42]:
# The statistics obtained are from the player's personal page
def getPersonalStats(playerLink, playerRecord):
      
  # Use playerLink to request a player's personal webpage
  playerPage = requests.get(playerLink)

  # This sleep is to make sure the sports reference site does not lock me oout while running as there is a limit of only 20 requests per minute can be done
  time.sleep(3.1)

  # Create a beautiful soup parser for player page
  playerSoup = BeautifulSoup(playerPage.text, 'html.parser')


  # Finding list of key value pairs that has player basic data in page as a list
  plist = playerSoup.find_all(text=re.compile("birth"))

  # Converting to actual python dictionary to parse through faster
  playerD = playerDictionaryConv(plist)
  # Grabbing the players birthday

  try:
    birthday = playerSoup.find(id = 'necro-birth')['data-birth']
  except:
    birthday = ''
  
  # Getting the rest of the stats from the individual player page
  getPersonalStatMapping(playerRecord, playerD, birthday, playerSoup)

  return playerRecord

# GetPersonalStat Interior Functions

In [43]:
# Turns re.compile list into a python dictionary
def playerDictionaryConv(string):
  
  # Ran into a weird edge case where some of the values returned divs from other parts fo the page so this removes them 
  string = [item for item in string if "div" not in item]


  stripped_string = ''.join(string).strip()

  #print(stripped_string)
  
  # Parse the JSON object
  try:
    data = json.loads(stripped_string)
  except:
    data = None

  # Convert data to dictionary
  try:
    dictionary = dict(data)
  
  except:
    keys = ['draft_year', 'draft_round' , 'draft_pick', 'draft_team', 'player_name', 'player_url', 'position', 'draft_age','college', 'height', 'weight', 'birth_year', 'birth_month', 'birth_day', 'college_stats_url', 'birth_place', 'high_school', 'image_url']
    dictionary = {key: None for key in keys}

  return dictionary

In [44]:
def getPersonalStatMapping(playerRecord, playerD, birthday, wholePage):
  
  #
  try: 
    playerRecord['birth_place'] = playerD['birthPlace']
  except:
    playerRecord['birth_place'] = None

  #
  try:
    #
    links = wholePage.find_all("a", href=lambda href: href and href.startswith("/schools/high_schools.cgi?"))
    names = 'HS,'
    #
    for link in links:
      names = names + link.text + ','
    #
    playerRecord['high_school']= names[:-1]
  except IndexError:
    playerRecord['high_school'] = None

  #
  try:
    playerRecord['image_url'] = playerD['image']['contentUrl']
  except:
    playerRecord['image_url'] = None
  
  #
  try:
    playerRecord['height'] = playerD['height']['value']
  except:
    playerRecord['height'] = None
  
  #
  try:
    playerRecord['weight'] = playerD['weight']['value']
  except:
    playerRecord['weight'] = None

  #
  try:
    bList = birthday.split('-')
    playerRecord['birth_year'] = bList[0]
    playerRecord['birth_month'] = bList[1]
    playerRecord['birth_day'] = bList[2]
  except:
    playerRecord['birth_year'] = None
    playerRecord['birth_month'] = None
    playerRecord['birth_day'] = None

  return playerRecord

# Garbage?

In [45]:
def dataStatAttrsInsert1(playerPage, record): 

  playerDataStatList = ['draft_pick', 'player', 'pos', 'college_id', 'team', 'college_link', 'age']
      
  # Appending text value of each tag with appropriate data-stat attribute
  for name in playerDataStatList:
    if name == 'draft_pick':
      record['draft_pick'] = playerPage.find('td', {'data-stat': name}).get_text()

    elif name == 'player':
      record['player_name'] = playerPage.find('td', {'data-stat': name}).get_text()

    elif name == 'pos':
      record['position'] = playerPage.find('td', {'data-stat': name}).get_text()

    elif name == 'team':
      record['draft_team'] = playerPage.find('td', {'data-stat': name}).get_text()

    elif name == 'college_link':
      try:
        td = playerPage.find('td', {'data-stat': name})
        if td:
            a = td.find('a')
            if a:
                record['college_stats_url'] = a['href']
            else:
                record['college_stats_url'] = None
        else:
            record['college_stats_url'] = None
      except IndexError:
        record['college_stats_url'] = None

    #
    elif name == 'age':
      record['draft_age'] = playerPage.find('td', {'data-stat': name}).get_text()

    else:
      record['college'] = playerPage.find('td', {'data-stat': name}).get_text()

  return record

In [46]:
def dataStatAttrsInsert1(playerPage, record): 

  playerDataStatList = ['draft_pick', 'player', 'pos', 'college_id', 'team', 'college_link', 'age']
      
  # Appending text value of each tag with appropriate data-stat attribute
  for name in playerDataStatList:
    if name == 'draft_pick':
      try:
        record['draft_pick'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['draft_pick'] = None

    elif name == 'player':
      try:
        record['player_name'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['player_name'] = None

    elif name == 'pos':
      try:
        record['position'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['position'] = None

    elif name == 'team':
      try:
        record['draft_team'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['draft_team'] = None

    elif name == 'college_link':
      try:
        td = playerPage.find('td', {'data-stat': name})
        if td:
            a = td.find('a')
            if a:
                record['college_stats_url'] = a['href']
            else:
                record['college_stats_url'] = None
        else:
            record['college_stats_url'] = None
      except (AttributeError, TypeError):
        record['college_stats_url'] = None

    elif name == 'age':
      try:
        record['draft_age'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['draft_age'] = None

    else:
      try:
        record['college'] = playerPage.find('td', {'data-stat': name}).get_text()
      except AttributeError:
        record['college'] = None
  print(record)
  return record

In [47]:
# Use boolean indexing to find the rows where the value in "column_name" is equal to "xyz"
drop_index = playerDF[playerDF['player_name'] == 'None'].index

# Drop the rows from the DataFrame
playerDF.drop(drop_index, inplace=True)